### ISSUE TITLE
Batch normalization

### ISSUE THREAD

**Issue by vdumoulin** [mila-udem/blocks#509]
As presented in [this paper](http://arxiv.org/pdf/1502.03167v3.pdf).

It would make a great feature for fully-connected as well as convolutional nets.

I think it would make sense to implement it as a wrapper class around `Activation` subclasses.

I intend on using it pretty soon, so I'd volunteer to implement it. Any thoughts, @dwf @bartvm @rizar ?

**Comment by dwf** [comment]
Was just thinking the same thing. Will gladly review. After I go read the paper. :)

**Comment by rizar** [comment]
Same for me, put the paper on the reading list.

**Comment by vdumoulin** [comment]
Thinking of it, we could implement BN in the same way dropout is implemented, using variable replacement.

At training time, we need to replace selected variables by subgraphs that
* Standardize the variables along some specified axes
* Apply a learned linear transformation to the standardized variables along certain axes

At test time, we need two things:
* Compute population statistics
* Apply a linear transformation to the selected variables according to these population statistics and the linear transformation learned during training

**Comment by vdumoulin** [comment]
@dwf @rizar I started a WIP PR (#513) to show what I have in mind. It currently only addresses the train-time requirements.
